### ISSUE TITLE
add Euro sign to keymaps

### ISSUE THREAD

**Issue by stsp** [dosemu2/dosemu2#836]
As was pointed here:
https://github.com/stsp/dosemu2/issues/18#issuecomment-499549422
the Euro sign should be added to keymaps.

**Comment by andrewbird** [comment]
So I started to look at adding the Euro sign to some of the keymaps, starting with the UK one.

1. A strange thing is that the Euro now works in terminal `-t` and `-td` display modes, even though neither `etc/keymaps` or `src/base/kbd_unicode/keymaps.c` show it defined for UK. On the UK keyboard the Euro symbol is generated using Alt-GR + 4, if I switch to FR or DE keymap in linux, the Euro is shown correctly by pressing Alt-Gr + E, which again is correct. I've made no dosemu keymap changes so far.

2. As a side note I also tried this same test in sdl mode using my sdlttf-01 branch and there the Euro sign displayed correctly, again without dosemu keymap change to UK. Switching linux desktop to FR or DE, did not allow the Alt-Gr + E to work as expected. Would the pcf/bdf font that dosemu uses for X and sdl have the Euro symbol defined, and if not perhaps that's the difference between devel and my sdlttf-01 branch.

So are the keymap changes now not required, and moreover is there keymap code that is no longer required?

**Comment by andrewbird** [comment]
So I see now that that in SDL and X mode the keymap seems to be working fine for UK also. The wrong character is being displayed only because I don't have a cp858 font to load.

**Comment by andrewbird** [comment]
So with cpi installed, I can use the Euro symbol with UK keyboard just fine. Are the BDF fonts not being used now?

**Comment by stsp** [comment]
Set `$_layout` to the layout you want to test.

**Comment by stsp** [comment]
> Are the BDF fonts not being used now?

Set `$_X_font` to the font you want to test.

**Comment by andrewbird** [comment]
> Set $_layout to the layout you want to test.

I get the same effect with autodetection giving me `fr-latin1` secondary with xkb switch as when I specify `$_layout` to be `fr-latin1`, that is euro doesn't work. That's what I expect with unaltered dosemu keymap files, but why does the euro work with `uk` keymap when they don't have it defined either?

**Comment by stsp** [comment]
Are you looking into `key_map_uk[]` in keymaps.c?

**Comment by andrewbird** [comment]
Well the alt map `alt_map_uk[]` there

**Comment by stsp** [comment]
At least I didn't touch the keymap code
since the last time you said Euro didn't work.

**Comment by andrewbird** [comment]
magic!

**Comment by stsp** [comment]
Does it have something to do with
$_internal_char_set which I think you
didn't set before?

**Comment by andrewbird** [comment]
Yes that's it, I commented it out and now in X, sdl and term is shows '?', but is that just because it doesn't have the character for euro? I tried altering the map to put a plain character in there, and that only prints as a '?' too, so it's like the map is not changing anything. I'll try a full rebuild, just in case there something that isn't being rebuilt.

**Comment by andrewbird** [comment]
I altered the plain keymap for uk too
```c
diff --git a/src/base/kbd_unicode/keymaps.c b/src/base/kbd_unicode/keymaps.c
index 121862c09..b33c63a89 100644
--- a/src/base/kbd_unicode/keymaps.c
+++ b/src/base/kbd_unicode/keymaps.c
@@ -237,7 +237,7 @@ CONST t_keysym key_map_uk[] =
 {
   U_VOID, 27, '1', '2', '3', '4', '5', '6',
   '7', '8', '9', '0', '-', '=', 127, 9,
-  'q', 'w', 'e', 'r', 't', 'y', 'u', 'i',
+  'B', 'w', 'e', 'r', 't', 'y', 'u', 'i',
   'o', 'p', '[', ']', 13, U_VOID, 'a', 's',
   'd', 'f', 'g', 'h', 'j', 'k', 'l', ';',
   '\'', '`', U_VOID, '#', 'z', 'x', 'c', 'v',
```

and the key still prints `q` in X, sdl and term

Off to bed now, it's late here. Does altering the keymap change any keys for you?

**Comment by stsp** [comment]
> and the key still prints q in X, sdl and term

Well, that's not true. :)
If you set `$_layout = "uk"` and use X (dosemu -X),
then the change in keymap has an effect, I checked.

**Comment by andrewbird** [comment]
Does have an effect (with $_layout = "uk") if the X desktop primary keyboard is set to something other than uk?

**Comment by andrewbird** [comment]
Also if you change a letter on your 'ru' keymap, do you see its effect. It seems to me that I see different effects on uk keyboard to say fr-latin1, in that I need to add the Euro symbol to fr-latin1 but I don't to uk.

**Comment by stsp** [comment]
I have primary layour US.

**Comment by andrewbird** [comment]
then forget the question about 'ru'

**Comment by stsp** [comment]
I have secondary RU.

**Comment by stsp** [comment]
Were you able to reproduce the change with -X?

**Comment by andrewbird** [comment]
I see change (using euro symbol) on fr-latin, I'll recheck uk.

**Comment by andrewbird** [comment]
with this patch
```c
diff --git a/src/base/kbd_unicode/keymaps.c b/src/base/kbd_unicode/keymaps.c
index 121862c09..644a6af7e 100644
--- a/src/base/kbd_unicode/keymaps.c
+++ b/src/base/kbd_unicode/keymaps.c
@@ -237,7 +237,7 @@ CONST t_keysym key_map_uk[] =
 {
   U_VOID, 27, '1', '2', '3', '4', '5', '6',
   '7', '8', '9', '0', '-', '=', 127, 9,
-  'q', 'w', 'e', 'r', 't', 'y', 'u', 'i',
+  'b', 'w', 'e', 'r', 't', 'y', 'u', 'i',
   'o', 'p', '[', ']', 13, U_VOID, 'a', 's',
   'd', 'f', 'g', 'h', 'j', 'k', 'l', ';',
   '\'', '`', U_VOID, '#', 'z', 'x', 'c', 'v',
```
1. with uk keymap as primary xkb, and $_layout commented out `q` is still printed

2. with uk keymap as primary xkb, and $_layout set to 'uk' `b` is printed.

**Comment by andrewbird** [comment]
Anyway, at least I understand how to make it work so I can fix the keymaps.c. Am I right in saying that the old `etc/keymap/*` are no longer installed? Shall I delete each of them after comparing to those in keymaps.c?

**Comment by andrewbird** [comment]
Next question keymaps.c has a lot of private code points e.g. `0xef9c`, why doesn't it use the standard `0x00A3` (U_POUND_SIGN)?

**Comment by stsp** [comment]
I think keymap/* are for customization.
Eg someone may have a non-standard
layout, so he should be able to customize
it. It seems you can force the external
keytable this way:
`$_layout = "load uk"`.
I am not sure how good does that work
currently, but I can think of writing
only the customized keys to the external
file (filling most of the table with U_VOID),
and it will override the particular kays
from keymaps.c.

> 0x00A3 (U_POUND_SIGN)?

Yes, I am not sure why there are such
things. If pure unicode works, we can
use that I think.

**Comment by andrewbird** [comment]
So I have another oddity with french keyboard in that there's different behaviour between X and sdl / terminals.

With `$_layout` set to 'fr-latin1', and primary xkb is 'UK, secondary is 'FR'
  a) `-X` keyboard in DOS is french layout
  b) `-S` keyboard in DOS is uk layout
  c) `-t`  keyboard in DOS is uk layout
  d) `-td` keyboard in DOS is uk layout

**Comment by stsp** [comment]
Yes, in SDL we trust XKB fully.
This is the only solution I've found.
You can trust dosemu layout in SDL
mode, but the price will be too high:
no switch to the secondary layout
(other than with keyb.com).

In t/td mode things are much different.
`$_layout` there is used only to find the
scancode. So if you see the "uk" layout
in td mode when you set "fr-latin", then
it is only half-way "uk": the scancodes
will be from "fr-latin". Example:
You swap `q` and `1` keys in your custom
layout, then press `q`. You'll see `q` on
screen, but if you read scancode from
kbd port, it will look like the scan from
key `1`.

So basically every mode uses the keytable
in a very different way.

**Comment by stsp** [comment]
If you set `$_layout=""` for auto, then X
will behave similar to SDL (i.e. trust XKB).
Interestingly, dosemu1 had 2 entirely
different modes for X: X_keysym and
X_keycode. I dropped both, but the rough
equivalent of dosemu1's X_keysym mode
still seems to exist (commit ada7125d
even updates it and says "this is the right
way of doing things", but later it was disabled
nevertheless... I think because I simply could
not remember so many different input modes).
In either case, if you change keyb_X.c:135 to
`config.X_keycode = 0;`
then you'll get X to behave similar to t/td modes
so you can use that for testing.

**Comment by stsp** [comment]
We probably need to document these things
somewhere, as its quite easy to forget. So in
dosemu2:
- SDL mode, X_auto mode: fully trust XKB
- X_layout mode: fully trust keytable (similar to X_keycode of dosemu1,
    not used in dosemu2 by default, can only be activated by altering `$_layout`)
- t/td mode, disabled X_unicode mode: use keytable for scancode look-ups

In dosemu1:
- SDL, t/td mode: use keytable for scancode look-ups
- X_keycode: fully trust keytable
- X_keysym: some strange mode that partially trusts XKB
    and partially the keytable (modified symbols are getting "fixed up" with keytable,
    non-modified go as-is)

So from that table it looks like dosemu2 uses new input
mode for SDL/X, t/td didn't change, X_keysym replaced
with X_unicode and disabled, X_keycode disabled by
default but still possible to re-enable,

TODO: check if `$_layout = "load XX"` work, and if it
does, can you only define a few custom buttons in an
external table, having the rest taken from the built-in keymap.

**Comment by andrewbird** [comment]
> SDL mode, X_auto mode: fully trust XKB

Not true for SDL; it's only partial. What happens is that with $_layout unset, and autodetect UK first, then FR-LATIN1 as alternate, only the UK works properly. If I switch active keyboard on desktop to FR, then basic layout like azerty works instead of qwerty. It seems as though only the keys that originally had mappings under UK work but with their FR mappings after the switch to FR, and any other keys that had no UK mappings are just void under FR.  How is the alternate Dosemu autodetect keymap supposed to be activated?

**Comment by stsp** [comment]
Iirc there is no any autodetection done
under sdl.
The keys that stop working - do they
have a character symbols, and do they
have a preceedind dead key? I think
for symbolic keys we trust xkb and they
should work.

**Comment by stsp** [comment]
> no UK mappings are just void under FR.

This is a but I think.
Please try
```
diff --git a/src/base/kbd_unicode/serv_xlat.c b/src/base/kbd_unicode/serv_xlat.c
index 81a4329f2..bf54add5e 100644
--- a/src/base/kbd_unicode/serv_xlat.c
+++ b/src/base/kbd_unicode/serv_xlat.c
@@ -1628,7 +1628,7 @@ static void put_keynum(Boolean make, t_keynum key, t_unicode sym, struct keyboar
                t_keysym *ch = get_rule_ptr(key, state);
                /* only substitute symbols from printable space.
                 * 0xe000 and above are non-unicode (private to t_keysym).  */
-               if (*ch < 0xe000 && *ch != sym) {
+               if (*ch != sym) {
                        k_printf("replace char %x with %x\n", *ch, sym);
                        *ch = sym;
                }
```
I think the "printable" check should be different.

**Comment by andrewbird** [comment]
I think autodetection via X and reading the console keymap (if perms allow) is always done regardless of the display. Using `-S` this is printed
~~~
CONF: detected layout is "uk"                                                   
CONF: detected alternate layout: fr-latin1                                      
~~~

Using your patch and my euro one, things seem to work okay. I'll run with it for a while before claiming it's fixed, as keyboard stuff seems to be very subtle.
~~~
ajb@polly:/clients/common/dosemu2.git$ git diff
diff --git a/src/base/kbd_unicode/keymaps.c b/src/base/kbd_unicode/keymaps.c
index ea74e80db..9966441df 100644
--- a/src/base/kbd_unicode/keymaps.c
+++ b/src/base/kbd_unicode/keymaps.c
@@ -501,7 +501,7 @@ CONST t_keysym alt_map_fr_latin1[] =
 {
   U_VOID, U_VOID, U_VOID, '~', '#', '{', '[', '|',
   '`', '\\', '^', '@', ']', '}', U_VOID, U_VOID,
-  '@', U_VOID, U_VOID, U_VOID, U_VOID, U_VOID, U_VOID, U_VOID,
+  '@', U_VOID, U_EURO_SIGN, U_VOID, U_VOID, U_VOID, U_VOID, U_VOID,
   U_VOID, U_VOID, U_VOID, 0xefa4, 13, U_VOID, U_VOID, U_VOID,
   U_VOID, U_VOID, U_VOID, U_VOID, U_VOID, U_VOID, U_VOID, U_VOID,
   U_VOID, U_VOID, U_VOID, U_VOID, U_VOID, U_VOID, U_VOID, U_VOID,
diff --git a/src/base/kbd_unicode/serv_xlat.c b/src/base/kbd_unicode/serv_xlat.c
index 81a4329f2..99b172b52 100644
--- a/src/base/kbd_unicode/serv_xlat.c
+++ b/src/base/kbd_unicode/serv_xlat.c
@@ -1628,7 +1628,8 @@ static void put_keynum(Boolean make, t_keynum key, t_unicode sym, struct keyboar
                t_keysym *ch = get_rule_ptr(key, state);
                /* only substitute symbols from printable space.
                 * 0xe000 and above are non-unicode (private to t_keysym).  */
-               if (*ch < 0xe000 && *ch != sym) {
+//             if (*ch < 0xe000 && *ch != sym) {
+               if (*ch != sym) {
                        k_printf("replace char %x with %x\n", *ch, sym);
                        *ch = sym;
                }
~~~

It's odd though, seeing that my new char is Euro 0x20ac, I'd have thought that char would have been replaced anyway.

**Comment by andrewbird** [comment]
Perhaps the sense of the test was wrong, should it be testing the new char < 0xe000, not the old? For example:
```c
                if (sym < 0xe000 && *ch != sym) {
                        k_printf("replace char %x with %x\n", *ch, sym);
                        *ch = sym;
                }
```

I now see this in the log
~~~
replace char ffff with 20ac
~~~

**Comment by andrewbird** [comment]
Note: that log is from your patch, not my suggestion.

**Comment by stsp** [comment]
> I think autodetection via X and reading the console keymap (if perms allow) is
> always done regardless of the display.

Indeed, the X plugin is loaded by the
generic keyboard code just for that.
Perhaps we need to separate that
code into "X_detect" plugin to avoid
confusion.

> Using your patch and my euro one, things seem to work okay.
> I'll run with it for a while before claiming it's fixed, as keyboard stuff seems to be very subtle.

This patch is definitely a breaker.
It is only for you to test Euro.

> Perhaps the sense of the test was wrong, should it be testing the new char < 0xe000, not the old?

I very much suspect not.
I'll see if I can come up with something else.

**Comment by stsp** [comment]
> Perhaps we need to separate that

Done.

**Comment by stsp** [comment]
I've found that `Xkb_lookup_key()` was in
`keyb_X.c` by mistake - it is only used by
`keyb_X_keycode.c`. I now moved it.
This means the table of different kbd modes
I tried to document yesterday, was spelled
wrongly. Need to try again.

**Comment by stsp** [comment]
I also now removed the difference between
X_layout and dosemu1's X_keycode mode to
simplify things.
So another attempt to document input methods:

dosemu2:
- SDL mode, X_auto mode: fully trust XKB (currently not always true, to be investigated)
- X_keycode mode: fully trust keytable (not default, can only be activated by altering $_layout)
- t/td mode, disabled X_keysym mode: use keytable for scancode look-ups

In dosemu1:
- SDL, t/td mode: use keytable for scancode look-ups
- X_keycode: fully trust keytable
- X_keysym: some strange mode that partially trusts XKB
    and partially the keytable (modified symbols are getting "fixed up" with keytable,
    non-modified go as-is)

So from that table it looks like dosemu2 uses new input
mode for SDL/X, t/td didn't change, X_keysym disabled,
X_keycode disabled by default but still possible to re-enable,

**Comment by stsp** [comment]
It seems the problem is in mishandling of an
Alt modifier. I think it is now worked around.

**Comment by stsp** [comment]
So now we trust XKB module the Alt-modified map.

**Comment by andrewbird** [comment]
So with latest devel + my euro patches I'm seeing some problems.
1. with `-t` I see a crash
~~~
ajb@polly:/clients/common/dosemu2.git$ ./runstandalonetest.sh -t
ERROR: Your fluidsynth is too old and soundfonts not found
ERROR: Fault in dosemu code, in_dpmi=0
ERROR: cpu exception in dosemu code outside of VM86()!
sig: 11 trapno: 0x0d  errorcode: 0x00000000  cr2: 0x00000000
eip: 0x563510aa088e  esp: 0x7ffd967937b0  eflags: 0x00010206
cs: 0x0033  ds: 0x0000  es: 0x0000  ss: 0x002b
fs: 0x0000  gs: 0x0000
fsbase: 0x7f42386a1780 gsbase: (nil)

ERROR: Please install gdb, update dosemu from git, compile it with debug
info and make a bug report with the content of ~/.dosemu/boot.log at
https://github.com/stsp/dosemu2/issues
Please provide any additional info you can, like the test-cases,
URLs and all the rest that fits.

mhpdbg.c:107:4: runtime error: null pointer passed as argument 1, which is declared to never be null
mhpdbg.c:109:4: runtime error: null pointer passed as argument 1, which is declared to never be null
~~~
[termcrash.log.gz](https://github.com/stsp/dosemu2/files/4144726/termcrash.log.gz)

2. SDL and X with $_layout unset no longer obey the fr-latin1 to display euro with key `Alt-Gr + e`

**Comment by stsp** [comment]
> with -t I see a crash

Should now be fixed.

> SDL and X with $_layout unset no longer obey the fr-latin1 to display euro with key Alt-Gr + e

Good to know, then its easier to finally
add it to a keymap. :)

**Comment by andrewbird** [comment]
I already added euro to my local branch for fr-latin1,  with it now rebased on your latest devel it no longer works. Sorry.

**Comment by stsp** [comment]
With explicit layout under X, or both ways?

**Comment by stsp** [comment]
Could you please try again?

**Comment by andrewbird** [comment]
1. So with `$_layout` unset, xkb (uk, fr), active = uk
  a. dosemu -X # Alt-Gr + 4 generates euro
  b. dosemu -S # Alt-Gr + 4 generates euro
  c. dosemu -t  # Alt-Gr + 4 generates euro

2. So with `$_layout` unset, xkb (uk, fr), active = fr
  a. dosemu -X # Alt-Gr + e generates euro
  b. dosemu -S # Alt-Gr + e generates nothing, but Alt-Gr + 4 gives euro as per inactive uk keymap
  c. dosemu -t  # Alt-Gr + e generates euro

3. So with `$_layout` == uk, xkb (uk, fr), active = uk
   a. dosemu - X # exit with `dosemu.bin: translate.c:633: init_charset_state: Assertion 'chars && state' failed.`
   b. dosemu -S # Alt-Gr + 4 generates euro
   c. dosemu -t  # Alt-Gr + 4 generates euro

4. So with `$_layout` == fr-latin, xkb (uk, fr), active = uk
  a. dosemu - X # exit with `dosemu.bin: translate.c:633: init_charset_state: Assertion 'chars && state' failed.`
  b. dosemu -S # Alt-Gr + e generates euro
  c. dosemu -t # Alt-Gr + e generates nothing, but Alt-Gr + 4 gives euro as per active uk keymap

So there seem to be problems in the cases (2b, 3a, 4a, 4c)

**Comment by stsp** [comment]
You mean with your euro patch?
Does it ever generate euro w/o any
patches?

**Comment by stsp** [comment]
Crash should now be fixed.
2b will require a bit of work.
4c will likely have to stay as is.

**Comment by andrewbird** [comment]
> You mean with your euro patch?

Yes

> Does it ever generate euro w/o any
patches?

I don't know at the moment, but right back at the start (before all your recent changes) euro was working without my euro patch for uk, but not fr. I'm a little uncertain which mode that was, but if I had to guess it would be sdl  with $_layout unset, xkb (uk, fr), active = uk.

> Crash should now be fixed.

I'll try again later.

**Comment by andrewbird** [comment]
1.    So with $_layout unset, xkb (uk, fr), active = uk
    a. dosemu -X # Alt-Gr + 4 generates euro
    b. dosemu -S # Alt-Gr + 4 generates euro
    c. dosemu -t # Alt-Gr + 4 generates euro

2.    So with $_layout unset, xkb (uk, fr), active = fr
    a. dosemu -X # Alt-Gr + e generates euro
    b. dosemu -S # Alt-Gr + e generates nothing, but Alt-Gr + 4 gives euro as per inactive uk keymap
    c. dosemu -t # Alt-Gr + e generates euro

3.    So with $_layout == uk, xkb (uk, fr), active = uk
    a. dosemu - X  # Alt-Gr + 4 generates euro
    b. dosemu -S  # Alt-Gr + 4 generates euro
    c. dosemu -t  # Alt-Gr + 4 generates euro

4.    So with $_layout == fr-latin, xkb (uk, fr), active = uk
    a. dosemu - X # Alt-Gr + e generates euro
    b. dosemu -S # Alt-Gr + e generates euro
    c. dosemu -t # Alt-Gr + e generates nothing, but Alt-Gr + 4 gives euro as per inactive uk keymap

So remaining problems are (2b, 4c)

**Comment by stsp** [comment]
Thanks, sounds better.

**Comment by stsp** [comment]
I think I've fixed the Euro also for X,default_layout,no_euro_patch scenario.
Please see if it is the case.

**Comment by andrewbird** [comment]
Cases 1b / 2b would be my, and I suspect others', usual ones.

**Comment by andrewbird** [comment]
> I think I've fixed the Euro also for X,default_layout,no_euro_patch scenario.

So you just want me to test pure devel?

**Comment by andrewbird** [comment]
Testing with latest devel, no euro patches.

1. So with $_layout unset, xkb (uk, fr, de), active = uk
   a. dosemu -X # Alt-Gr + 4 generates euro
2. So with $_layout unset, xkb (uk, fr, de), active = fr
   a. dosemu -X # Alt-Gr + e generates euro
3. So with $_layout unset, xkb (uk, fr, de), active = de
   a. dosemu -X # Alt-Gr + e generates euro

Note that de works, even though it is not autodetected by Dosemu as primary or alternate keymap

**Comment by andrewbird** [comment]
There is slight oddity, in that sometimes the first keypress of an `alt-gr + some key` produces the bare character without the alt map applied, but subsequent keypresses work normally.

**Comment by andrewbird** [comment]
Also, in the case that DE is XKB active, but is neither the Dosemu primary nor alternate, the Ctrl-Alt-PgDn doesn't work. Maybe that's a step too far, but I thought I'd mention it!

**Comment by stsp** [comment]
Is the altgr oddity related to the last
patch?

**Comment by andrewbird** [comment]
I think so, but can't be sure
 1. I just tested HEAD~1 with no euro patch, with X on uk xkb (uk, fr), uk active, `Alt-Gr + 4` printed '$' consistently. Strangely, '$' is in the shifted map, but there was no hint of printing '4'.
 2. I just tested HEAD with no euro patch, with X on xkb (uk, fr), uk active, `Alt-Gr + 4` printed '€' consistently.
 3. I just tested HEAD with no euro patch, with X on xkb (uk, fr), fr active, `Alt-Gr + e` printed nothing on first try, then '€'

**Comment by stsp** [comment]
Would be nice to test 3 on HEAD~1

**Comment by andrewbird** [comment]
I just tested HEAD~1 with no euro patch, with X on xkb (uk, fr), fr active, Alt-Gr + e printed nothing (but that's expected as it has U_VOID in that position in the fr-latin1 alt map) Pressing `Alt-Gr + a` printed '@' consistently with no misses.

**Comment by stsp** [comment]
So w/o last patch there is no inconsistent
behaviour where repeating the keypress
gives something else?

**Comment by andrewbird** [comment]
> So w/o last patch there is no inconsistent
behaviour where repeating the keypress
gives something else?

That's correct, but the keys generally produce nothing as the alt_map is mostly empty. And with the latest patch in my observation it's always the initial keypress that is wrong, subsequent ones are as expected.

**Comment by stsp** [comment]
Could you please debug this a little?
There is a condition:
```
	if (state & (X_mi.AltMask | X_mi.AltGrMask))
		return DKY_VOID;
```

Is this condition differs for the first press,
or maybe it just never causes the premature
return, and the problem happens later?
What if you just comment both lines?

**Comment by andrewbird** [comment]
I added this instrumentation
```diff
ajb@polly:/clients/common/dosemu2.git$ git diff
diff --git a/src/plugin/X/keyb_X_keycode.c b/src/plugin/X/keyb_X_keycode.c
index 16cea81d9..1cbcb80b4 100644
--- a/src/plugin/X/keyb_X_keycode.c
+++ b/src/plugin/X/keyb_X_keycode.c
@@ -629,8 +629,12 @@ static t_unicode Xkb_lookup_key(Display *display, KeyCode keycode,
 
        /* XXX alt is not represented in a sym and doesn't return error
         * from XkbTranslateKeySym(), so disable it by hands. :( */
-       if (state & (X_mi.AltMask | X_mi.AltGrMask))
+       if (state & (X_mi.AltMask | X_mi.AltGrMask)) {
+               k_printf("AJB - returned DKY_VOID state == 0x%x\n", state);
                return DKY_VOID;
+       } else {
+               k_printf("AJB - no return state == 0x%x\n", state);
+       }
        /* XXX Ctrl-Enter seems to be misconfigured:
         * https://github.com/stsp/dosemu2/issues/864
         * Disable it for now. */
```
I just tested HEAD with only this patch, with X on xkb (uk, fr), fr active, Alt-Gr + e printed nothing on first try, then '€'

Here's the log [test.log.gz](https://github.com/stsp/dosemu2/files/4155063/test.log.gz)

**Comment by andrewbird** [comment]
And commenting out both lines still resulted in `Alt-Gr + e` printed nothing on first try, then '€'.

**Comment by andrewbird** [comment]
Is it possible that `X_get_modifier_info()` is returning out of date info, as it seems just to return the contents of a static var?

**Comment by stsp** [comment]
> Is it possible that X_get_modifier_info() is returning out of date info

Its only used for masks.
"state" is a current modifier info.

**Comment by andrewbird** [comment]
So X_modifier_info_init() sets the value of X_mi just once, does it never change?

**Comment by stsp** [comment]
Its just a modifier masks.
Modifier masks (AltMask, AltGrMask) obviously
never change. "state" represents the run-time
modifiers.

**Comment by stsp** [comment]
I think that should now be fixed.
Please check.

**Comment by andrewbird** [comment]
Just devel, no Euro patches of mine applied.

1.    So with $_layout unset, xkb (uk, fr), active = uk
    a. dosemu -X # Alt-Gr + 4 generates euro
    b. dosemu -S # Alt-Gr + 4 generates '$'
    c. dosemu -t # Alt-Gr + 4 generates euro

2.    So with $_layout unset, xkb (uk, fr), active = fr
    a. dosemu -X # Alt-Gr + e generates euro
    b. dosemu -S # Alt-Gr + e generates nothing, but Alt-Gr + 4 gives '$' as per inactive uk keymap
    c. dosemu -t # Alt-Gr + e generates euro

3.    So with $_layout == uk, xkb (uk, fr), active = uk
    a. dosemu -X # Alt-Gr + 4 generates '$'
    b. dosemu -S # Alt-Gr + 4 generates '$'
    c. dosemu -t # Alt-Gr + 4 generates euro

4.    So with $_layout == fr-latin1, xkb (uk, fr), active = uk
    a. dosemu -X # Alt-Gr + e generates nothing (no euro on Alt-Gr + 4 as per active uk keymap)
    b. dosemu -S # Alt-Gr + e generates nothing (no euro on Alt-Gr + 4 as per active uk keymap)
    c. dosemu -t # Alt-Gr + e generates 'e', but Alt-Gr + 4 gives euro as per active uk keymap

So problems are 1b, 2b, 3a, 3b, 4a, 4b, 4c , but maybe some will be helped now by my euro keymap patches?

**Comment by andrewbird** [comment]
With devel + my two euro keymap patches

 1.   So with $_layout unset, xkb (uk, fr), active = uk
    a. dosemu -X # Alt-Gr + 4 generates euro
    b. dosemu -S # Alt-Gr + 4 generates euro
    c. dosemu -t # Alt-Gr + 4 generates euro

2.    So with $_layout unset, xkb (uk, fr), active = fr
    a. dosemu -X # Alt-Gr + e generates euro
    b. dosemu -S # Alt-Gr + e generates nothing, but Alt-Gr + 4 gives euro as per inactive uk keymap
    c. dosemu -t # Alt-Gr + e generates euro

3.    So with $_layout == uk, xkb (uk, fr), active = uk
    a. dosemu -X # Alt-Gr + 4 generates euro
    b. dosemu -S # Alt-Gr + 4 generates euro
    c. dosemu -t # Alt-Gr + 4 generates euro

4.    So with $_layout == fr-latin1, xkb (uk, fr), active = uk
    a. dosemu -X # Alt-Gr + e generates euro
    b. dosemu -S # Alt-Gr + e generates euro
    c. dosemu -t # Alt-Gr + e generates 'e', but Alt-Gr + 4 gives euro as per active uk keymap

So problems are 2b, 4c

**Comment by stsp** [comment]
Yes, that looks much better.
With your patch we have -X flawless
and -S quite close to.
Thanks.

**Comment by stsp** [comment]
So now I suppose we need to fix 2b
and close this.

**Comment by andrewbird** [comment]
Yes, fixing 2b would be excellent.

**Comment by stsp** [comment]
There is actually more than that.
For example you can't specify multiple
layouts, like `$_layout="us,ru"`.
But I don't want to dive into the large
rewrite.
2b needs a small rewrite though, so
I'll try to get to that.

**Comment by stsp** [comment]
I created a "kbd_map" patch where I
tried to address the problem. But I do not
have a test-case, and am unsatisfied
with the patch intrusiveness.
Does it work?

**Comment by andrewbird** [comment]
I'm assuming this is for the 2b problem, is that correct? If so I don't think it works as it seems to be the same as before.

1. So with $_layout unset, xkb (uk, fr), active = uk
b. dosemu -S # Alt-Gr + 4 generates euro

2. So with $_layout unset, xkb (uk, fr), active = fr
b. dosemu -S # Alt-Gr + e generates nothing, but Alt-Gr + 4 gives euro as per inactive uk keymap


For the next 2 weeks or so I'll be travelling, with only access to a laptop with US/Chinese keyboard(long story!), so I won't be able to test these euro things promptly.

**Comment by andrewbird** [comment]
Scratch that, it's in a branch, I'll retest!

**Comment by andrewbird** [comment]
No, not right for uk or fr now.

1. So with $_layout unset, xkb (uk, fr), active = uk
b. dosemu -S # Alt-Gr + 4 generates '$'

2. So with $_layout unset, xkb (uk, fr), active = fr
b. dosemu -S # Alt-Gr + e generates nothing

**Comment by stsp** [comment]
It doesn't have your updates applied.

**Comment by andrewbird** [comment]
okay, rebuilding with cherry-picked uk and fr patches

**Comment by andrewbird** [comment]
1.    With $_layout unset, xkb (uk, fr), active = uk
    b. dosemu -S # Alt-Gr + 4 generates euro

2.    With $_layout unset, xkb (uk, fr), active = fr
    b. dosemu -S # Alt-Gr + e euro

So that looks good, I didn't get the chance to do any further testing than those two cases. As I mentioned before I'll be away from this keyboard for a couple of weeks, so won't be able to test further.

**Comment by stsp** [comment]
Thanks!
That patch likely breaks -X slightly,
and overall needs more thinking.
But the fact that it works, is encouraging.

**Comment by stsp** [comment]
I think it is there now?
